src/bootstrap/format.rs

```diff
@@ -193,10 +193,46 @@ pub fn format(build: &Builder<'_>, check: bool, paths: &[PathBuf]) {
     let (tx, rx): (SyncSender<PathBuf>, _) = std::sync::mpsc::sync_channel(128);
     let walker = match paths.get(0) {
         Some(first) => {
-            let mut walker = WalkBuilder::new(first);
+            let find_shortcut_candidates = |p: &PathBuf| {
+                let mut candidates = Vec::new();
+                for candidate in WalkBuilder::new(src.clone()).max_depth(Some(3)).build() {
+                    if let Ok(entry) = candidate {
+                        if let Some(dir_name) = p.file_name() {
+                            if entry.path().is_dir() && entry.file_name() == dir_name {
+                                candidates.push(entry.into_path());
+                            }
+                        }
+                    }
+                }
+                candidates
+            };
+
+            // Only try to look for shortcut candidates for single component paths like
+            // `std` and not for e.g. relative paths like `../library/std`.
+            let should_look_for_shortcut_dir = |p: &PathBuf| p.components().count() == 1;
+
+            let mut walker = if should_look_for_shortcut_dir(first) {
+                if let [single_candidate] = &find_shortcut_candidates(first)[..] {
```

Comment:
Are you sure this is right? Does it still work for `x fmt library`?

Reply:
I can share a simple demonstration.

<details>
<summary>Current changes</summary>

```diff
diff --git a/library/std/benches/hash/map.rs b/library/std/benches/hash/map.rs
index bf646cbae47..30834887db6 100644
--- a/library/std/benches/hash/map.rs
+++ b/library/std/benches/hash/map.rs
@@ -13,7 +13,7 @@ fn new_drop(b: &mut Bencher) {

 #[bench]
 fn new_insert_drop(b: &mut Bencher) {
-    b.iter(|| {
+            b.iter(|| {
         let mut m = HashMap::new();
         m.insert(0, 0);
         assert_eq!(m.len(), 1);
diff --git a/library/std/src/backtrace/tests.rs b/library/std/src/backtrace/tests.rs
index 4dfbf88e83e..f3563806043 100644
--- a/library/std/src/backtrace/tests.rs
+++ b/library/std/src/backtrace/tests.rs
@@ -6,7 +6,7 @@ fn generate_fake_frames() -> Vec<BacktraceFrame> {
             frame: RawFrame::Fake,
             symbols: vec![BacktraceSymbol {
                 name: Some(b"std::backtrace::Backtrace::create".to_vec()),
-                filename: Some(BytesOrWide::Bytes(b"rust/backtrace.rs".to_vec())),
+                        filename: Some(BytesOrWide::Bytes(b"rust/backtrace.rs".to_vec())),
                 lineno: Some(100),
                 colno: None,
             }],
```

</details>

<details>
<summary>fmt check output</summary>

```sh
  ~/devspace/personal/rust2  remotes/jieyouxu/issue-107944 $ ./x fmt library --check
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.05s
Diff in /home/nimda/devspace/personal/rust2/library/std/benches/hash/map.rs at line 13:

 #[bench]
 fn new_insert_drop(b: &mut Bencher) {
-            b.iter(|| {
+    b.iter(|| {
         let mut m = HashMap::new();
         m.insert(0, 0);
         assert_eq!(m.len(), 1);
Diff in /home/nimda/devspace/personal/rust2/library/alloc/src/boxed/thin.rs at line 64:
     #[cfg(not(no_global_oom_handling))]
     pub fn new(value: T) -> Self {
         let meta = ptr::metadata(&value);
-                    let ptr = WithOpaqueHeader::new(meta, value);
+        let ptr = WithOpaqueHeader::new(meta, value);
         ThinBox { ptr, _marker: PhantomData }
     }
 }
Running `"/home/nimda/devspace/personal/rust2/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/home/nimda/devspace/personal/rust2" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/home/nimda/devspace/personal/rust2/library/std/benches/hash/set_ops.rs" "/home/nimda/devspace/personal/rust2/library/std/benches/hash/mod.rs" "/home/nimda/devspace/personal/rust2/library/std/benches/hash/map.rs" "/home/nimda/devspace/personal/rust2/library/std/benches/lib.rs" "/home/nimda/devspace/personal/rust2/library/std/build.rs" "/home/nimda/devspace/personal/rust2/library/std/src/sys/wasm/atomics/thread.rs" "/home/nimda/devspace/personal/rust2/library/std/src/macros.rs" "/home/nimda/devspace/personal/rust2/library/panic_unwind/src/hermit.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
Build completed unsuccessfully in 0:00:01
```

</details>

<details>
<summary>fmt and git status</summary>

```sh
  ~/devspace/personal/rust2  remotes/jieyouxu/issue-107944 $ ./x fmt library
Building bootstrap
    Finished dev [unoptimized] target(s) in 0.07s
Build completed successfully in 0:00:01
  ~/devspace/personal/rust2  remotes/jieyouxu/issue-107944 $ git status
HEAD detached from e7293ca5a9f
nothing to commit, working tree clean
  ~/devspace/personal/rust2  remotes/jieyouxu/issue-107944 $
```

</details>

Reply:
Ok great :) I guess `find_shortcut_candidates` is doing the right thing, thanks for testing.